Jetty-12 Remove usage of HandlerList and reduce usage of Handler.Collection (#9191)
eclipse/jetty.project: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/http/server-http-handler-use.adoc

Server applications may need to deploy to Jetty more than one web application.

> Recall from the xref:pg-server-http-handler[introduction] that Jetty offers `Handler.Collection` that may contain a collection of children ``Handler``s.

Reviewer: ```suggestion
Recall from the xref:pg-server-http-handler[introduction] that Jetty offers `Handler.Collection` that contains a sequence of child ``Handler``s.
```

However, this has no knowledge of the concept of _context_ and just iterates through the sequence of ``Handler``s.

A better choice for multiple web application is `ContextHandlerCollection`, that matches a _context_ from either its _context path_ or _virtual host_, without iterating through the ``Handler``s.

> If `ContextHandlerCollection` does not find a match, it just returns false from ``process``.

Reviewer 2: ```suggestion
If `ContextHandlerCollection` does not find a match, it just returns `false` from its `process(\...)` method.
```

What happens next depends on the `Handler` tree structure: other ``Handler``s may be invoked after `ContextHandlerCollection`, for example `DefaultHandler` (see xref:pg-server-http-handler-use-util-default-handler[this section]).

> Eventually, if no ``Handler`` returns `true` from `process`, then Jetty returns an HTTP `404` response to the client.

Reviewer 2: ```suggestion
Eventually, if no `Handler` returns `true` from their own `process(\...)` method, then Jetty returns an HTTP `404` response to the client.
```

include::../../{doc_code}/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java[tags=multipleResourcesHandler]
----

> If the resource is not found, `ResourceHandler` will not return `true` from the `process` method, so what happens next depends on the `Handler` tree structure.

Reviewer 2: ```suggestion
If the resource is not found, `ResourceHandler` will not return `true` from the `process(\...)` method, so what happens next depends on the `Handler` tree structure.
```

[[pg-server-http-handler-use-util-default-handler]]
====== DefaultHandler

> `DefaultHandler` is a terminal `Handler` that always returns `true` from its `process` method and performs the following:

Reviewer 2: ```suggestion
`DefaultHandler` is a terminal `Handler` that always returns `true` from its `process(\...)` method and performs the following:
```

When a request arrives to `ServletContextHandler` the request URI will be matched against the ``Filter``s and ``Servlet`` mappings and only those that match will process the request, as dictated by the Servlet specification.

> IMPORTANT: `ServletContextHandler` is a terminal `Handler`, that is it always returns `true` from its `process` method when invoked.

Reviewer 2: ```suggestion
IMPORTANT: `ServletContextHandler` is a terminal `Handler`, that is it always returns `true` from its `process(\...)` method when invoked.
```

> Server applications must be careful when creating the `Handler` tree to put ``ServletContextHandler``s as last ``Handler``s in a `HandlerList` or as children of `ContextHandlerCollection`.

Reviewer 2: Still a reference to `HandlerList` here?
